Improve how "sqlfluff fix" handles templating and parse errors

barrywhart opened the pull request:
- [x] Fixing stdin should support same or similar behavior
- [x] Config file setting `fix_even_unparsable`
- [x] Simple API should support same or similar behavior as command line?
- [x] Multi-file tests
- [x] Documentation

<!--Thanks for adding this feature!-->

<!--Please give the Pull Request a meaningful title for the release notes-->

### Brief summary of the change made
Fixes #2345, #2424, #2251, #2134

### Are there any other side effects of this change that we should be aware of?


### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

tunetheweb commented on docs/source/production.rst:
```diff
@@ -106,6 +106,18 @@ SQLFluff comes with two `pre-commit`_ hooks:
+   For safety reasons, ``sqlfluff-fix`` by default will not make any fixes in
+   files that had parse errors.
+
+   Although it is not advised, you _can_ tell SQLFluff to try and fix
```
What does the underline mean? Should it be in italics?

```suggestion
   Although it is not advised, you `can` tell SQLFluff to try and fix
```

barrywhart replied:
Oh yeah. Markdown confusion. Will fix.

tunetheweb commented on docs/source/production.rst:
```diff
@@ -106,6 +106,18 @@ SQLFluff comes with two `pre-commit`_ hooks:
+   files that had parse errors.
+
+   Although it is not advised, you _can_ tell SQLFluff to try and fix
+   unparseable files by overriding the ``fix_even_unparsable`` setting
+   (``.sqlfluff`` or ``sqlfluff fix`` command line).
```
```suggestion
   unparseable files by overriding the ``fix_even_unparsable`` setting
   in ``.sqlfluff`` config file or using the ``sqlfluff fix --FIX-EVEN-UNPARSABLE``
   command line option.
```

tunetheweb commented on docs/source/production.rst:
```diff
@@ -106,6 +106,18 @@ SQLFluff comes with two `pre-commit`_ hooks:
 * sqlfluff-fix: attempts to fix rule violations.
 
+.. note::
+   For safety reasons, ``sqlfluff-fix`` by default will not make any fixes in
+   files that had parse errors.
```
Is it same for templating errors? Or just parse errors?

barrywhart replied:
Thus far, I've only touched the behavior for parse errors. Let me think about and experiment with templating errors. They should probably behave the same way.

barrywhart replied:
Ok, I updated it to handle templater errors the same way. Need to review a little more closely to see if there's leftover code to clean up, but it should be working correctly and consistently now.

tunetheweb commented on src/sqlfluff/api/simple.py:
```diff
@@ -123,8 +123,17 @@ def fix(
     result = linter.lint_string_wrapped(sql, fix=True)
-    fixed_string = result.paths[0].files[0].fix_string()[0]
-    return fixed_string
+    fix_even_unparsable = cfg.get("fix_even_unparsable")
+    should_fix = fix_even_unparsable
+    if not should_fix:
```
I'm very confused by this.

1) Why do we need two variables (`fix_even_unparsable` and `should_fix`)?
2) `if not should_fix` is doing my head in. Any way to rephrase without the `not`? Especially as it seems to be a variable created just for this.

barrywhart replied:
It's because there are not one but *two* ways that `should_fix` can be set to `True`, either initially, based on `fix_even_unparsable`, or inside the `if` statement. Basically, we "should fix" the file if either:
* `fix_even_unparsable` is set **OR**
* The file has no parse errors. The code inside the `if` checks for this.

There are other ways to structure this code that may be clearer, but I think they would involve duplicating some of the code.

tunetheweb replied:
Well the way it’s currently written is a bit of an optimisation - it only calls the unparsable check if it needs to. So those using the force option (that we want to discourage) save that check.

Given that we want to discourage that force nature, I’m not really inclined to optimise for them, and would favour code readability over that.

So I would say do the check and then after, if no parse errors OR flag is set, do the fix.

WDYT?

barrywhart replied:
I agree the current code looks odd when you initially see it, but I'm also trying to consider if you're a maintainer and you take a closer look, you're likely to ask, "Why is it doing all this useless work to count errors when it doesn't need to? Maybe I'm missing something. I'm afraid to touch this code."

I'm making a slight change that you may like better.

tunetheweb commented on src/sqlfluff/api/simple.py:
```diff
@@ -123,8 +123,17 @@ def fix(
+        # If fix_even_unparsable wasn't set, check for parse errors. Okay to fix
+        # if there are no parse errors.
+        _, num_filtered_parse_errors = result.check_parse_errors()
+        if not num_filtered_parse_errors:
+            should_fix = True
```
Would this not be more readible stated as this?:

```suggestion
        if num_filtered_parse_errors == 0:
            should_fix = True
```

barrywhart replied:
Works for me!

tunetheweb commented on src/sqlfluff/core/default_config.cfg:
```diff
@@ -27,6 +27,11 @@ disable_noqa = False
 # Comma separated list of file extensions to lint
 # NB: This config will only apply in the root folder
 sql_file_exts = .sql,.sql.j2,.dml,.ddl
+# By default, files won't be fixed if they contain parse errors (even if those
+# errors are ignored with "noqa"). This option allows overriding that behavior:
+# if set to True, fixes *will* be applied. USE WITH CAUTION: MAY APPLY SPURIOUS
+# OR UNNECESSARY FIXES. IT MAY BREAK OR CORRUPT YOUR SQL.
+fix_even_unparsable = False
```
Could we make this smaller?

```suggestion
# Allow fix to run on files, even if they contain parsing errors
# Note altering this is NOT RECOMMENDED as can corrupt SQL
fix_even_unparsable = False
```

barrywhart replied:
I often wonder how much we need to scare people. We'll go with the shorter phrasing for now. We can make it scarier later if people keep shooting themselves in the foot and reporting issues. 😅

tunetheweb replied:
I was thinking the big block made it stand out (which we don’t really want)! 😁 So tried to minimise it a bit.

tunetheweb commented on test/cli/commands_test.py:
```diff
@@ -509,6 +511,149 @@ def test__cli__command__fix(rule, fname):
+            ],
+            ["--force", "--fixed-suffix", "FIXED", "--rules", "L010"],
+            [
+                None,
```
Should this be None? Or should it be the original sql?

barrywhart replied:
`None` is a special value: It means that `fix` did nothing, i.e. it didn't even _write_ an output file. Technically, they are almost the same, but if we're using `--fixed-suffix`, then `None` makes it clearer that the "fixed" file shouldn't even be created. Same contents would suggest the file was written, but without changes. See the test body and you'll see what I mean -- with `None`, it literally checks that the output file is not present.

tunetheweb replied:
Grand. I just presumed that fix returned the original SQL if nothing to fix. Think the simple API does that. But probably makes more sense there. Anyway must be right as tests pass! So was more a question for my understanding.

barrywhart commented on docs/source/configuration.rst:
```diff
@@ -644,6 +644,7 @@ You might have arguments that you pass through every time, e.g rules you
     [sqlfluff]
     verbose = 1
     exclude_rules = L022,L027
+    fix_even_unparsable = True
```
I was unsure. Having it here makes people aware that it can be configured, but if we don't explain that it's a bad idea, it may encourage people to use it inappropriately. I'll remove it.

tunetheweb commented on src/sqlfluff/api/simple.py:
```diff
@@ -123,8 +123,17 @@ def fix(
+        # if there are no parse errors.
+        _, num_filtered_parse_errors = result.check_parse_errors()
+        if num_filtered_parse_errors == 0:
+            should_fix = True
```
Doesn’t this need to be `!=` and also setting it to false?

btw do like this better. Though personally think I would have gone for parse error check first and then see if needed. Even if it’s a slight waste of processing for those using the force flag.

barrywhart replied:
Ah yes. Cuz it's already set to true. The risk of refactoring working code. 😉

tunetheweb commented on docs/source/production.rst:
```diff
@@ -106,6 +106,19 @@ SQLFluff comes with two `pre-commit`_ hooks:
 * sqlfluff-lint: returns linting errors.
 * sqlfluff-fix: attempts to fix rule violations.
 
+.. note::
```
```suggestion
.. warning::
```

barrywhart commented:
I'm putting this back in draft mode because I "discovered" the existing `--ignore templating,parsing` command-line option. I think we should use that, rather than adding a new `--FIX-EVEN-UNPARSABLE` option.

@tunetheweb, @alanmcruickshank

tunetheweb commented:
> I'm putting this back in draft mode because I "discovered" the existing `--ignore templating,parsing` command-line option. I think we should use that, rather than adding a new `--FIX-EVEN-UNPARSABLE` option.

That’s a slightly different thing though. It basically means ignore them from linting. Your own means “fix what you can even if linting errors, whether ignored or not”.

barrywhart commented:
>That’s a slightly different thing though. It basically means ignore them from linting.
>Your own means “fix what you can even if linting errors, whether ignored or not”.

Are you sure? Look at [these messages](https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/cli/commands.py#L672L685) printed in `fix` when there are templater errors.  Specifically, this one:
>Use '--ignore templating' to attempt to fix anyway.

🤯

barrywhart commented:
@tunetheweb: Oh yes, you're right. `--ignore` is almost identical to `noqa`, but applied globally. 👍

barrywhart commented on src/sqlfluff/core/linter/linting_result.py:
```diff
@@ -181,3 +184,34 @@ def tree(self) -> Optional[BaseSegment]:  # pragma: no cover
+                            if hasattr(violation, "fixes"):
+                                violation.fixes = []  # type: ignore
+                        num_filtered_errors += linted_file.num_violations(types=types)
+        return total_errors, num_filtered_errors
```
This function is the main part of the PR. Every flavor of "fix" eventually uses it:
* `fix` command line
* `fix` command line for stdin (special case)
* Simple API `fix()` function

barrywhart commented on test/cli/commands_test.py:
```diff
@@ -509,6 +511,162 @@ def test__cli__command__fix(rule, fname):
+            ],
+            1,
+        ),
+    ],
```
These test cases are worth a careful review. They should match with the behavior we discussed in the maintainers meeting and added to issue #2424:

`sqlfluff fix`:
- Only fix if no unparsable segments found in the file. If any are found, no fixes happen even if parsing errors are supressed using noqa or `--ignore` cli options. The only way to override this is with the `--FIX-EVEN-UNPARSABLE` option as above.
- If there are still parse errors after applying "noqa" directives, exit 1. Otherwise, allow exit 0.

alanmcruickshank replied:
I think these test cases reflect what we intended 👍 .

alanmcruickshank commented on src/sqlfluff/api/simple.py:
```diff
@@ -123,8 +123,17 @@ def fix(
     result = linter.lint_string_wrapped(sql, fix=True)
-    fixed_string = result.paths[0].files[0].fix_string()[0]
-    return fixed_string
+    fix_even_unparsable = cfg.get("fix_even_unparsable")
```
For API usage, I think this should be an available kwarg on `fix()`, and not just accessible via the config.

barrywhart replied:
Done.

alanmcruickshank commented on src/sqlfluff/core/linter/linting_result.py:
```diff
@@ -181,3 +184,34 @@ def tree(self) -> Optional[BaseSegment]:  # pragma: no cover
             )
         return self.paths[0].tree
+
+    def check_templating_or_parse_errors(self) -> Tuple[int, int]:
+        """Treat tiles with templating or parse error as unfixable.
+
+        NOTE! THIS FUNCTION HAS SIDE EFFECTS: IT DISCARDS LINT FIXES.
+
```
I can see why this function has side effects. Ideally it wouldn't, but I think they way this function is used it's more of a pruning function or filter function on the current set of violations.

Given that it *does* have side effects however, I wonder if we should change the name. `check_<x>` makes it sound like a function which _doesn't_ have side effects.

barrywhart replied:
Done.

alanmcruickshank reviewed (changes requested):
Two minor changes - but the bulk of this is spot on I think 👍

tunetheweb commented on src/sqlfluff/cli/commands.py:
```diff
@@ -678,11 +716,13 @@ def fix(
             click.echo(
                 colorize(
-                    "Use '--ignore templating' to attempt to fix anyway.",
+                    "Use '--ignore templating' and '--fix-even-unparsable' to "
+                    "attempt to fix anyway.",
```
Do you need both? I thought you only needed the latter and the former just controlled the exit code? Though not sure how to succinctly put that!

What about:

```suggestion
                    "Use --fix-even-unparsable' to attempt to fix the SQL anyway, ",
                    " and use '--ignore templating' to avoid outputing templating ",
                    " errors as well.",
```

barrywhart replied:
Will do.

barrywhart replied:
Technically, you're absolutely right that only `--fix-even-unparsable` is now needed. I was unsure of the best wording here, and thought that users may want to be aware of both options. It's a good point, though. I'll remove mention of `--ignore templating`.

barrywhart replied:
Done.

tunetheweb commented on src/sqlfluff/core/linter/linting_result.py:
```diff
@@ -181,3 +184,32 @@ def tree(self) -> Optional[BaseSegment]:  # pragma: no cover
             )
         return self.paths[0].tree
+
+    def mark_failed_files_unfixable(self) -> Tuple[int, int]:
+        """Discard lint fixes for files with templating or parse errors.
```
The function name and the description differs.

What about:

```suggestion
    def discard_unfixable_errors(self) -> Tuple[int, int]:
        """Discard lint fixes for files with templating or parse errors.
```

barrywhart replied:
It's a difficult function to name. The name you proposed is also inaccurate: we're not discarding _errors_, we're discarding _fixes_ for errors and keeping the actual errors.

How about `discard_fixes_for_lint_errors_in_files_with_templating_or_parsing_errors`?

barrywhart replied:
Renamed as `discard_fixes_for_lint_errors_in_files_with_tmp_or_prs_errors()`.

tunetheweb replied:
A function name almost as long as your branch names 😁

tunetheweb commented on src/sqlfluff/core/linter/linting_result.py:
```diff
@@ -181,3 +184,32 @@ def tree(self) -> Optional[BaseSegment]:  # pragma: no cover
+
+        Scan all LintedFiles:
+        - Files with none of these errors: No action
+        - Files with these errors (before filtering): Removes fixes from
+          SQLLintError objects, i.e. marks them "unfixable"
```
I misread this that we are removing `SQLLintError`s. But actually what we're doing is removing `SQLParseError` and `SQLTemplaterError` from the list of linting errors and leaving the other `SQLLintError`s still. Maybe this could be clearer?

```suggestion
        - Files with these errors (before filtering): Removes fixes for
          `SQLParseError` and `SQLTemplaterError` errors (i.e. marks them
          "unfixable").
```

barrywhart replied:
It's not removing fixes for SQLParseError and SQLTemplaterError errors. It's removing fixes for SQLLintError in *files* with SQLParseError and SQLTemplaterError. The SQLParseError and SQLTemplaterError objects are left untouched.

barrywhart replied:
I'm changing the `hasattr(violation, "fixes")` to `isinstance(violation, SQLLintError)`. This will make it clearer what's happening. 👍

tunetheweb commented on src/sqlfluff/core/linter/linting_result.py:
```diff
@@ -181,3 +184,32 @@ def tree(self) -> Optional[BaseSegment]:  # pragma: no cover
+        - Files with these errors (before filtering): Removes fixes from
+          SQLLintError objects, i.e. marks them "unfixable"
+
+        Also returns the total number of these errors before and after
+        "ignore/noqa" filters are applied.
```
I wonder if the number of errors after "ignore/noqa" filters are applied should be part of this function?

Seems separate (and short) so could be pulled out.

Just concerned there's potential for confusion between the removal (aka "filtering") of TMP/PRS errors, and the filtering of `ignore`/`--noqa` errors.

barrywhart replied:
Done.

tunetheweb reviewed (commented):
Sorry, but lots of comments and questions.

It works and so happy with code from that perspective, but think it could be clearer to read, for future.

barrywhart commented on src/sqlfluff/cli/commands.py:
```diff
@@ -575,6 +576,28 @@ def lint(
         sys.exit(0)
 
 
+def _mark_failed_files_unfixable(lint_result: LintingResult) -> int:
+    """Discard lint fixes for files with templating or parse errors.
+
+    Returns 1 if there are any files with templating or parse errors after
+    filtering, else 0. (Intended as a process exit code.)
+    """
+    total_errors, num_filtered_errors = lint_result.mark_failed_files_unfixable()
```
One of them is a wrapper around the other, and they're in different modules. There's a similar (albeit exact) duplication of function names across the linter modules and classes, because of the layered structure of the linter (commands -> LintingResult -> LintedDir -> LintedFile:
* `get_violations()`
* `num_violations()`
* `persist_changes()`
* etc.

barrywhart replied:
Renamed both these functions. The names are more distinct now.

barrywhart commented on src/sqlfluff/cli/commands.py:
```diff
@@ -575,6 +576,28 @@ def lint(
+        click.echo(
+            colorize(f"  [{total_errors} templating/parsing errors found]", Color.red)
+        )
+        if 0 < num_filtered_errors < total_errors:
```
>What if there are no filtered errors at all but there are TMP/PRS errors?
`num_filtered_errors` is a variable that counts the number of filtered TMP and PRS errors, so I don't understand the question.

>Should it be 0 <= num_filtered_errors < ...? At which point do you even need this and should it just be num_filtered_errors < total_errors?
I originally didn't have the `0 <` part. I added it because I was thinking, if *all* the errors are filtered, why print an "error" message saying "0 errors after filtering", because that's _good_, right?

barrywhart replied:
I removed the `0 < ` from the `if`, so now the `no cover` shouldn't be needed, because existing tests will now hit this code. Instead, I changed it so the color of the output is red if there are still errors after filtering, green if there are none.

barrywhart commented on src/sqlfluff/cli/commands.py:
```diff
@@ -575,6 +576,28 @@ def lint(
+                    f'remaining after "ignore"]',
+                    Color.red,
+                )
+            )  # pragma: no cover
```
As noted above, I changed the `if` statement a bit. Now this line will be executed by the existing tests.

tunetheweb replied:
Fair point about it just being output btw. Overlooked that. But good that it's not needed anyway now.

barrywhart commented:
@tunetheweb: Okay, I think I've addressed all (almost all?) of your feedback. Ready for another review. I removed the `#pragma: no cover` from the `click.echo()` after changing the code around it, so I think it'll be covered by existing tests. We'll know for sure once the build finishes. 🍿

tunetheweb commented on src/sqlfluff/api/simple.py:
```diff
@@ -123,8 +124,19 @@ def fix(
+    if fix_even_unparsable is None:
+        fix_even_unparsable = cfg.get("fix_even_unparsable")
+    should_fix = True
+    if not fix_even_unparsable:
+        # If fix_even_unparsable wasn't set, check for templating or parse
+        # errors and suppress fixing if there were any.
+        _, num_filtered_errors = result.count_tmp_prs_errors()
+        result.discard_fixes_for_lint_errors_in_files_with_tmp_or_prs_errors()
+        if num_filtered_errors > 0:
+            should_fix = False
+    if should_fix:
+        sql = result.paths[0].files[0].fix_string()[0]
```
Sorry, I'm still having trouble with this (head now hurts!)

Why do we need to call `discard_fixes_for_lint_errors_in_files_with_tmp_or_prs_errors` here now?

If there are PRS or TMP errors (whether `noqa`ed or not) then we will set `should_fix = False` so won't fix, so no reason to discard.
If there are no PRS or TMP errors, then no reason to discard.

barrywhart replied:
We don't. This is an optimization exposed by splitting the one function into two functions. Will change.

barrywhart replied:
Done!

barrywhart commented:
>A function name almost as long as your branch names 😁
I was trying to avoid it, but it's hard to explain in a few words. 😛  Next time, `magic_code_xxx123()`. 🎉

tunetheweb reviewed (approved):
I think we got there!

barrywhart commented:
@tunetheweb: Whee! Onward and upward. 🚀
